### PR TITLE
[BYOC] Added "include_non_call_ops" parameter to AnnotateTarget pass

### PR DESCRIPTION
**Pull request by d-smirnov**
This PR adds "include_non_call_ops" parameter to AnnotateTarget pass. When set (include_non_call_ops=True) the AnnotateTarget pass will annotate non call ops with default target or with the target of its arguments. This is current behavior of AnnotateTarget pass.  When the flag is set to false (include_non_call_ops=False) the AnnotateTarget pass will not annotate non-call. This behavior is useful if you are not running MergeCompilerRegions pass after AnnotateTarget.

The PR related to an issue reported here: https://discuss.tvm.apache.org/t/arm-compute-library-segv-with-inception-v1-squeezenet/7985/8

@comaniac @manupa-arm @mbaret

### CODE REVIEW

**Comment by comaniac** [comment]
Thanks for the PR. I have a hard deadline this weekend and will review it after next Tuesday.

File: src/relay/transforms/annotate_target.cc
Comment (manupa-arm) [inline src/relay/transforms/annotate_target.cc:83]: Maybe its worth adding a comment, this is to not touch (non-empty and non-default) annotation. Took sometime figure out the unusual compiler_begins append at this location of the code. Just a suggestion.

File: src/relay/transforms/annotate_target.cc
Comment (manupa-arm) [reply src/relay/transforms/annotate_target.cc:128]: This leaves the above comment out of place; Now it seems we only bypass compiler begins if its default or empty. Thus, move the comment here and change accordingly. Anyway, is there a reason why we are bypassing "default" compiler_begins here and adding them back in AnnotateArgs ? -- If we did not bypass them, we dont need to add them back (in line 80) ?

File: src/relay/transforms/annotate_target.cc
Comment (manupa-arm) [inline src/relay/transforms/annotate_target.cc:139]: Same question here as well; reason for treating already put "default" annotation as they were empty ?

File: tests/python/relay/test_pass_annotate_target.py
Comment (manupa-arm) [inline tests/python/relay/test_pass_annotate_target.py:362]: While the number of partitions gives an indication whether tuples are going to into partition, don't you think checking explicitly for sub-graphs of tuples make sense?

**Review by manupa-arm — CHANGES_REQUESTED** [review]
Thanks for this! some minor comments.

File: python/tvm/relay/transform/transform.py
Comment (trevor-m) [inline python/tvm/relay/transform/transform.py:669]: Perhaps `include_tuples=True` is more descriptive?

File: tests/python/relay/test_pass_annotate_target.py
Comment (mbaret) [inline tests/python/relay/test_pass_annotate_target.py:359]: I think we should be testing the failure case we saw, which was when MergeCompilerRegions wasn't run.

**Review by mbaret — CHANGES_REQUESTED** [review]

File: tests/python/relay/test_pass_annotate_target.py
Comment (manupa-arm) [reply tests/python/relay/test_pass_annotate_target.py:358]: There is the test_multiple_runs() there, I think its run multiple times and checks for equivalency as if it ran a single time using two targets.

Comment (mbaret) [reply tests/python/relay/test_pass_annotate_target.py:358]: Oh OK, makes me wonder why we're carrying the complexity of allowing multiple targets then, but that's a question for another time...

Comment (comaniac) [reply tests/python/relay/test_pass_annotate_target.py:358]: @mbaret see the description in https://github.com/apache/incubator-tvm/pull/5277. I recall this is the requirement from you or the other ARM folk. In that case you need to annotate differernt targets before and after QNN passes.

Comment (mbaret) [reply tests/python/relay/test_pass_annotate_target.py:358]: Yeah, it's just confusing that we have two different ways to achieve the same result. Given multiple runs are supported, it seems like we don't need to complexity of supporting a list of targets.

Comment (comaniac) [reply tests/python/relay/test_pass_annotate_target.py:358]: List of targets serves for the different purpose. List of target allows us to have a whole picture during the annotation. We can determine which target an op should be annotated by considering the region size, for example. As a result, I personally prefer to have a list of targets instead of multiple runs if we could. However, given the scenario you provided, which two targets require different graphs, it seems inevitable.

**Comment by comaniac** [comment]
This solution looks too ad hoc to me. From the semantic of the changes, I think this PR can be generalized to "improve the annotate target pass to annotate non-call ops to default". Accordingly, the API could be like

```python
AnnotateTarget(targets, skip_non_call_ops=False)(mod) # I am bad at naming but it is the point
```

cc @zhiics

File: python/tvm/relay/transform/transform.py
Comment (zhiics) [inline python/tvm/relay/transform/transform.py:669]: This API sounds quite ad-hoc to me. `Tuple` is just one of the nodes that can take the same target from its inputs/args. Incrementally adding parameters to handle different nodes would make the API bulky and code less readable, IMHO.

File: src/relay/transforms/annotate_target.cc
Comment (d-smirnov) [reply src/relay/transforms/annotate_target.cc:80]: Done

File: src/relay/transforms/annotate_target.cc
Comment (d-smirnov) [reply src/relay/transforms/annotate_target.cc:43]: Done

File: python/tvm/relay/transform/transform.py
Comment (d-smirnov) [reply python/tvm/relay/transform/transform.py:669]: Could you elaborate the list of non-call nodes that should have this feature along with the Tuple, please?

Comment (manupa-arm) [reply python/tvm/relay/transform/transform.py:669]: I think what @zhiics means is all the overridden [Rewrite_]s of annotate_target except CallNode -- where Tuple is just one of them.

Comment (d-smirnov) [reply python/tvm/relay/transform/transform.py:669]: For example, current implementation of `Let` in relay does not accept anything other than `Var` as its first parameter. Effectively, it cannot be annotated with anything but `"default"`. So for `Let` these changes would not make any difference.

Comment (manupa-arm) [reply python/tvm/relay/transform/transform.py:669]: I think if skip_non_call_ops=True, we should not annotate Lets -- even with defaults. Thus, the behavior is not the same.

**Comment by d-smirnov** [comment]
@tqchen Hi, I am trying to compile Mobilnet (after expanded this patch to all non-call ops, not committed yet) and encountered a check on src/relay/backend/compile_engine.cc:170
`ICHECK(op->is_scalar());`

The ConstantNode it fails is of TensorType: TensorType([1, 1, 1024, 1024], float32), and contains runtime.NDArray(0x74f0370) of corresponding shape (1x1x1024x1024). Could you explain the purpose of the check and should the check be extended to also accomodate ConstantNodes of TensorType. I am also interested to know the techniques to trace when and why this ConstantNode appeared. 

Thank you. cc: @comaniac

**Comment by trevor-m** [comment]
> @tqchen Hi, I am trying to compile Mobilnet (after expanded this patch to all non-call ops, not committed yet) and encountered a check on src/relay/backend/compile_engine.cc:170
> `ICHECK(op->is_scalar());`
> 
> The ConstantNode it fails is of TensorType: TensorType([1, 1, 1024, 1024], float32), and contains runtime.NDArray(0x74f0370) of corresponding shape (1x1x1024x1024). Could you explain the purpose of the check and should the check be extended to also accomodate ConstantNodes of TensorType. I am also interested to know the techniques to trace when and why this ConstantNode appeared.
> 
> Thank you. cc: @comaniac

Hey @d-smirnov have you tried with this change? https://github.com/apache/incubator-tvm/pull/6912

**Comment by d-smirnov** [comment]
> Hey @d-smirnov have you tried with this change? #6912
Rebased, but it did not make any difference. Still fail on the check `src/relay/backend/compile_engine.cc:170`

**Comment by d-smirnov** [comment]
Updated. Please take a look @comaniac, @mbaret, @manupa-arm

File: src/relay/transforms/annotate_target.cc
Comment (comaniac) [inline src/relay/transforms/annotate_target.cc:121]: This makes L188 unreachable?

File: src/relay/transforms/annotate_target.cc
Comment (comaniac) [inline src/relay/transforms/annotate_target.cc:149]: No effect?

File: src/relay/transforms/annotate_target.cc
Comment (comaniac) [inline src/relay/transforms/annotate_target.cc:148]: Cannot connect this comment to the following logic. Could you elaborate?

File: src/relay/transforms/annotate_target.cc
Comment (comaniac) [inline src/relay/transforms/annotate_target.cc:191]: Looks like you don't need the IF? Even `input_expr` is already in `op_expr_to_target_`, you can still override it, as suggested by the comment in L188. Accordingly, if you will override the target, you need `InsertAnnotation`.

File: src/relay/transforms/annotate_target.cc
Comment (comaniac) [inline src/relay/transforms/annotate_target.cc:205]: Looks like you remove the feature that considers the target in existing annotation nodes?

File: src/relay/transforms/annotate_target.cc
Comment (comaniac) [inline src/relay/transforms/annotate_target.cc:248]: Could you elaborate why a CallNode may have a VarNode as its op?

**Review by comaniac — CHANGES_REQUESTED** [review]
I still think we should have separate postprocess logic to deal with the subgraph (a.k.a. region) without call nodes instead of coupling it to AnnotateTarget (similar to the pruning pass in TensorRT integration). The AnnotateTarget pass is already too complicate to maintain, and that's not what we expected. On the other hand, we can maybe let this in first and plan another refactor.

In addition, since this PR may change the output of AnnotateTaret, you need to add corresponding tests to the follow-up passes (i.e., MergeCompilerRegion and GraphPartition).

@rohanmukh @codeislife99 please help check if this change affect any existing workloads.

cc @zhiics @anijain2305

File: tests/python/relay/test_pass_annotate_target.py
Comment (d-smirnov) [reply tests/python/relay/test_pass_annotate_target.py:797]: fixed

File: tests/python/relay/test_pass_annotate_target.py
Comment (d-smirnov) [reply tests/python/relay/test_pass_annotate_target.py]: removed

File: python/tvm/relay/transform/transform.py
Comment (d-smirnov) [reply python/tvm/relay/transform/transform.py]: refactored as include_non_call_ops, defaults to True

File: src/relay/transforms/annotate_target.cc
Comment (d-smirnov) [reply src/relay/transforms/annotate_target.cc]: Refactored

File: src/relay/transforms/annotate_target.cc
Comment (d-smirnov) [reply src/relay/transforms/annotate_target.cc]: refactored

File: src/relay/transforms/annotate_target.cc
Comment (d-smirnov) [reply src/relay/transforms/annotate_target.cc]: refactored

File: src/relay/transforms/annotate_target.cc
Comment (d-smirnov) [reply src/relay/transforms/annotate_target.cc:86]: It removes default (but preserve non-default) annotations in case if it is not first invocation of the pass

File: src/relay/transforms/annotate_target.cc
Comment (d-smirnov) [reply src/relay/transforms/annotate_target.cc:149]: Removed.

File: src/relay/transforms/annotate_target.cc
Comment (d-smirnov) [reply src/relay/transforms/annotate_target.cc:148]: The comment related to this part `(call && !call->args.empty()))` of the condition

File: src/relay/transforms/annotate_target.cc
Comment (d-smirnov) [reply src/relay/transforms/annotate_target.cc:205]: Here it is peeking first arg and if it is already annotated with non-default target it returns the node untouched, preserving the target.

File: src/relay/transforms/annotate_target.cc
Comment (d-smirnov) [reply src/relay/transforms/annotate_target.cc:248]: The test case is tests/python/relay/test_pass_annotate_target.py::test_while_let

File: src/relay/transforms/annotate_target.cc
Comment (d-smirnov) [reply src/relay/transforms/annotate_target.cc:191]: Removed IF

File: src/relay/transforms/annotate_target.cc
Comment (d-smirnov) [reply src/relay/transforms/annotate_target.cc:121]: I do not understand you comment. please elaborate.

File: tests/python/relay/test_pass_annotate_target.py
Comment (d-smirnov) [reply tests/python/relay/test_pass_annotate_target.py:217]: Done. Removed

File: src/relay/transforms/annotate_target.cc
Comment (comaniac) [reply src/relay/transforms/annotate_target.cc:121]: nvm. I figured it out.

File: src/relay/transforms/annotate_target.cc
Comment (comaniac) [reply src/relay/transforms/annotate_target.cc:86]: My question was about the algorithm. I don't understand why we "remove default (but preserve non-default) annotations in case if it is not first invocation of the pass". Almost all my comments related to this question.

Meanwhile, PR derives another AnnotateTarget implementation to deal with the case that `include_non_call_ops=False`, so you are more like making a new pass and use `include_non_call_ops` to dispatch the one being used. Combining my previous concern about keeping AnnotateTarget simple and straightforward, I would suggest making it a separate pass. For example, we could name the new pass "PruneSimpleRegions" and use it as following:

1. (MergeComposiste) -> AnnotateTarget -> PruneSimpleRegions -> PartitionGraph
2. (MergeComposiste) -> AnnotateTarget -> MergeCompilerRegions -> PruneSimpleRegions -> PartitionGraph

The first case could achieve your goal, and it could also serve the second case to prune subgraphs (compiler regions) after merging. If we take one step further, we could even let users provide a lambda function to determine if a region is "simple", so that it can be customized for different backends such as TensorRT. The implementation of this pass could also be much simpler and I could even help with that if you prefer.

What do you think?

Comment (d-smirnov) [reply src/relay/transforms/annotate_target.cc:86]: Current contract for AnnotateTarget assumes that for non-call op to be promoted under the target of its arguments it should have all the arguments annotated with the same target. If I am not mistaken, the targets can be added at the run-time, and theoretically the situation is possible when after several subsequent invocations of AnnotateTarget a non-call operation will be annotated as default and will not be promoted even when all its arguments have same target. To prevent this the default annotations always removed while specific targets are preserved. However I am agree that the explanation is a bit far-fetched.

As for the second part of your comment, do I understand correctly that the idea is not to alter the AnnotateTarget and instead do the removal of annotations from non-call ops in a subsequent separate pass? I am not sure I quite understand the definition of "simple" and the concept of PruneSimpleRegions as a whole. Could you elaborate this point a bit more, please?

Comment (comaniac) [reply src/relay/transforms/annotate_target.cc:86]: Thanks for the explanation. I think I am now clearer about the first part, but we might need to improve the code comments a lot to make it clear to everyone reading this pass.

For the second part, your understanding is correct. As you mentioned, the definition of "simple" is vague, but it's also more general. Since every target needs to specify their own transform pipeline (e.g., https://github.com/apache/incubator-tvm/blob/main/python/tvm/relay/op/contrib/arm_compute_lib.py#L44), we can every BYOC target can define their own "simple". For example, the definition of "simple" in ACL could be a subgraph with non-call ops; the definition of "simple" in TensorRT could be a subgraph without MAC ops. It means something like

```python
    seq = tvm.transform.Sequential(
        [
            transform.InferType(),
            transform.MergeComposite(arm_compute_lib_pattern_table()),
            transform.AnnotateTarget("arm_compute_lib"),
            transform.PruneSimpleRegions(tvm._ffi.get_global_func("relay.ext.arm_compute_lib.prune")),
            transform.PartitionGraph(),
        ]
    )
```

where `tvm._ffi.get_global_func("relay.ext.arm_compute_lib.prune")` is a packed function in C++ that accepts an `AnnotatedRegion` and outputs a boolean, indicating whether the region should be pruned or not.

Comment (d-smirnov) [reply src/relay/transforms/annotate_target.cc:86]: The proposal is likely to be more flexible. But I am afraid that introducing an externally defined method for doing pruning also means introducing a more elaborate checker for IR as it should be clearly stated what is the correctly pruned graph and what is not.

Comment (comaniac) [reply src/relay/transforms/annotate_target.cc:86]: IMHO, it should be fine, because we have abstracted the region to a separate data structure called "AnnotatedRegion". If we pass an AnnotatedRegion to PruneSimpleRegions, we just work on the region level. We can add an API such as "ChangeRegionTarget" to AnnotatedRegion that changes all compiler attributes of the nodes in that region. In this case, PruneSimpleRegion could just do (in Pseudo code):

```
for region in regions:
    if region.target != "default" and is_simple_region(region):
        region.change_region_target("default")
```

and one implementation of "is_simple_region" coud be like:

```
def is_simple_region(region):
    class Checker(ExprVisitor):
        def __init__(self):
            self.is_simple = True
        def visit(self, call):
            self.is_simple = False
            super(self).visit(call)

    checker = Checker()
    for out in region.outs:
        checker.visit(out)
    return checker.is_simple
```

Comment (mbaret) [reply src/relay/transforms/annotate_target.cc:86]: I think we need to separate the (valid) concerns about the design and overall maintainability of AnnotateTarget from the intention of this PR. Currently, we have user facing errors with ACL because of this tuple annotation bug and so I think our first priority should be merging something which fixes that. We can discuss possible refactors/redesigns of AnnotateTarget separately with an RFC on the forum where more people can see it and participate.

File: src/relay/transforms/annotate_target.cc
Comment (zhiics) [inline src/relay/transforms/annotate_target.cc:150]: There would be more nodes, like constructors. But I am still concerned if this changed is needed. This really makes this already complicated pass more complicated. I still don't see a good point why we don't run mergecompilerregions. That would solve this problem. Without running it, we would have a large number of small segments, which requires frequent data transfer between the host and device as well as frequent kernel launch.

Comment (manupa-arm) [reply src/relay/transforms/annotate_target.cc:150]: While running merge compiler regions helps cut down the regions, it also makes the external codegen's responsibility to allocate memory for intermediate tensors on those partitions. Thus, in the specific case of ACL, I think there is not much gained by such a merger as ACL would be implementing each ACL primitive operator and let tvm handle the memory allocation of the tensors passed onto external function. Moreover, the kernel launch overhead  should also be minimal as it is running on the CPU (so the host and device is almost the same here). Also such a merger will also make the IO tensors live throughout the execution of external function while the space could be re-used if it was not merged.

The problem is the specification of the ACL did not indicate the simple regions (or non-call ops) to be annotated, thus annotate target here is doing something extra than it was asked. 

I quite agree that this pass is complicated and needs breakdown. I guess that should be discussed in a RFC as to how it should look like. One direction maybe to take out the annotation of simple regions (non-call ops) as a seperate part ( I believe this was how it looked liked sometime back when it had something AnnotateRestDefault until it got merged here :) ).

Comment (manupa-arm) [reply src/relay/transforms/annotate_target.cc:150]: See my comment here : https://github.com/apache/tvm/pull/5277

File: tests/python/relay/test_pass_annotate_target.py
Comment (d-smirnov) [reply tests/python/relay/test_pass_annotate_target.py]: this is outdated, hence closing

File: src/relay/transforms/annotate_target.cc
Comment (d-smirnov) [reply src/relay/transforms/annotate_target.cc:86]: @comaniac Bump

Comment (comaniac) [reply src/relay/transforms/annotate_target.cc:86]: Aggregate the discussions to https://github.com/apache/tvm/pull/6655#pullrequestreview-538712328

File: src/relay/transforms/annotate_target.cc
Comment (zhiics) [reply src/relay/transforms/annotate_target.cc:150]: yeah, it would be nice to have the RFC and list the options there

**Comment by d-smirnov** [comment]
@comaniac  Ping. How can we make some progress here?

**Comment by comaniac** [comment]
> @comaniac Ping. How can we make some progress here?

I still don't think we should make AnnotateTarget pass even more complicate, so I proposed to have a separate ACL specific pass to workaround this issue like TensorRT. Later we could have an RFC to aggregate those passes and make a single, general pass for all BOYC targets.

cc @zhiics @mbaret @manupa-arm for comments.

**Comment by mbaret** [comment]
I don't agree that this should be separate, because the current behaviour of AnnotateTarget is simply incorrect. ACL does not support tuples, so AnnotateTarget should not mark them as supported. We shouldn't need a 'fix-up' pass after AnnotateTarget to make it correct, it's just a bug in AnnotateTarget that's come about from faulty reasoning about how codegens work. As this is user-facing and a critical error, I think the priority is accepting a fix with a refactor to reduce complexity coming later.

**Review by comaniac — APPROVED** [review]
Per offline discussion, we will let this PR in first. Afterwards, @mbaret will send an RFC discussing the desire behavior of AnnotateTarget.

@mbaret please merge this PR if that works for you. Thanks.

**Review by mbaret — APPROVED** [review]
LGTM, but yes we need to specify the expected behaviour of this pass more formally. We'll follow up with an RFC to this effect in the new year.

**Comment by mbaret** [comment]
Thanks everyone.

**Pull request merged**
